# Project 1 - digits (손글씨 분류)
## module import

In [1]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

## 데이터 준비

In [2]:
digits = load_digits()

In [3]:
print(dir(digits))
print("digits keys : ",digits.keys())

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']
digits keys :  dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])


### Feature Data 지정하기

In [4]:
# data의 shape
digits_data = digits.data
print(digits_data.shape)

(1797, 64)


In [5]:
digits_data[0]

array([ 0.,  0.,  5., 13.,  9.,  1.,  0.,  0.,  0.,  0., 13., 15., 10.,
       15.,  5.,  0.,  0.,  3., 15.,  2.,  0., 11.,  8.,  0.,  0.,  4.,
       12.,  0.,  0.,  8.,  8.,  0.,  0.,  5.,  8.,  0.,  0.,  9.,  8.,
        0.,  0.,  4., 11.,  0.,  1., 12.,  7.,  0.,  0.,  2., 14.,  5.,
       10., 12.,  0.,  0.,  0.,  0.,  6., 13., 10.,  0.,  0.,  0.])

### Label Data 지정하기

In [6]:
# label data 지정 
digits_label = digits.target
print(digits_label.shape)

(1797,)


### Target Names 출력해 보기

In [7]:
# target names
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

### 데이터 Describe 해 보기

In [8]:
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 5620
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [9]:
# feature names
digits.feature_names

['pixel_0_0',
 'pixel_0_1',
 'pixel_0_2',
 'pixel_0_3',
 'pixel_0_4',
 'pixel_0_5',
 'pixel_0_6',
 'pixel_0_7',
 'pixel_1_0',
 'pixel_1_1',
 'pixel_1_2',
 'pixel_1_3',
 'pixel_1_4',
 'pixel_1_5',
 'pixel_1_6',
 'pixel_1_7',
 'pixel_2_0',
 'pixel_2_1',
 'pixel_2_2',
 'pixel_2_3',
 'pixel_2_4',
 'pixel_2_5',
 'pixel_2_6',
 'pixel_2_7',
 'pixel_3_0',
 'pixel_3_1',
 'pixel_3_2',
 'pixel_3_3',
 'pixel_3_4',
 'pixel_3_5',
 'pixel_3_6',
 'pixel_3_7',
 'pixel_4_0',
 'pixel_4_1',
 'pixel_4_2',
 'pixel_4_3',
 'pixel_4_4',
 'pixel_4_5',
 'pixel_4_6',
 'pixel_4_7',
 'pixel_5_0',
 'pixel_5_1',
 'pixel_5_2',
 'pixel_5_3',
 'pixel_5_4',
 'pixel_5_5',
 'pixel_5_6',
 'pixel_5_7',
 'pixel_6_0',
 'pixel_6_1',
 'pixel_6_2',
 'pixel_6_3',
 'pixel_6_4',
 'pixel_6_5',
 'pixel_6_6',
 'pixel_6_7',
 'pixel_7_0',
 'pixel_7_1',
 'pixel_7_2',
 'pixel_7_3',
 'pixel_7_4',
 'pixel_7_5',
 'pixel_7_6',
 'pixel_7_7']

## dataframe으로 바꾸기

In [10]:
import pandas as pd

In [11]:
digits_df = pd.DataFrame(data = digits.data, columns = digits.feature_names)
digits_df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

## label 넣어주기

In [12]:
digits_df['label'] = digits.target

In [13]:
digits_df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_7  pixel_7_0  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        0.0        0.0   
3           0.0        0.0        0.0        8.0  ...        0.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        0.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        0.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        0.0        0.0   

      pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  pixel_7_6  \
0           0.0        6.0       13.0       10.0        0.0        0.0   
1           0.0        0.0       11.0       16.0       10.0        0.0   
2           0.0        0.0        3.0       11.0       16.0        9.0   
3           0.0        7.0       13.0       13.0        9.0        0.0   
4           0.0        0.0        2.0       16.0        4.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        2.0       14.0       15.0        9.0        0.0   
1793        0.0        6.0       16.0       14.0        6.0        0.0   
1794        0.0        2.0        9.0       13.0        6.0        0.0   
1795        0.0        5.0       12.0       16.0       12.0        0.0   
1796        1.0        8.0       12.0       14.0       12.0        1.0   

      pixel_7_7  label  
0           0.0      0  
1           0.0      1  
2           0.0      2  
3           0.0      3  
4           0.0      4  
...         ...    ...  
1792        0.0      9  
1793        0.0      0  
1794        0.0      8  
1795        0.0      9  
1796        0.0      8  

[1797 rows x 65 columns]

## train_set, test_set 분리

In [14]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(digits_data, digits_label, test_size=0.2, random_state=32)
print("X_train 개수: {0}, X_test 개수: {1}, y_train 개수: {2}, y_test 개수: {3}".format(len(X_train), len(X_test), len(y_train), len(y_test)))

X_train 개수: 1437, X_test 개수: 360, y_train 개수: 1437, y_test 개수: 360


X_train, test, y_train, test의 shape

In [15]:
print("X_train shape: {0}\nX_test shape: {1}\ny_train shape: {2}\ny_test shape: {3}".format(X_train.shape,X_test.shape,y_train.shape,y_test.shape))

X_train shape: (1437, 64)
X_test shape: (360, 64)
y_train shape: (1437,)
y_test shape: (360,)


## 모델정의

In [16]:
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(random_state=32)
print(model._estimator_type)

classifier


## 모델학습 및 예측

In [17]:
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.89      0.94        38
           1       0.75      0.83      0.79        36
           2       0.75      0.84      0.79        32
           3       0.92      0.86      0.89        56
           4       0.85      0.90      0.88        31
           5       0.92      0.97      0.95        36
           6       0.97      0.94      0.96        34
           7       0.94      0.88      0.91        34
           8       0.88      0.78      0.82        27
           9       0.79      0.83      0.81        36

    accuracy                           0.88       360
   macro avg       0.88      0.87      0.87       360
weighted avg       0.88      0.88      0.88       360



## 1. DecisionTreeClassifier로 예측 평가

In [18]:
# module import
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
digits=load_digits()
digits_data = digits.data
digits_label = digits.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data, digits_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = DecisionTreeClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.95      0.96        38
           1       0.76      0.78      0.77        36
           2       0.82      0.84      0.83        32
           3       0.94      0.88      0.91        56
           4       0.87      0.84      0.85        31
           5       0.81      0.97      0.89        36
           6       0.94      0.94      0.94        34
           7       0.91      0.88      0.90        34
           8       0.88      0.78      0.82        27
           9       0.81      0.83      0.82        36

    accuracy                           0.87       360
   macro avg       0.87      0.87      0.87       360
weighted avg       0.88      0.87      0.87       360



## 2. RandomForestClassifier로 예측 평가

In [19]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

# 데이터 준비하기
digits=load_digits()
digits_data = digits.data
digits_label = digits.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data, digits_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = RandomForestClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        38
           1       0.95      1.00      0.97        36
           2       1.00      1.00      1.00        32
           3       1.00      0.98      0.99        56
           4       1.00      0.97      0.98        31
           5       0.97      0.97      0.97        36
           6       1.00      1.00      1.00        34
           7       1.00      0.97      0.99        34
           8       0.93      0.93      0.93        27
           9       0.97      1.00      0.99        36

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



## 3. SVM으로 예측평가

In [20]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import classification_report

# 데이터 준비하기
digits=load_digits()
digits_data = digits.data
digits_label = digits.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data, digits_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = svm.SVC()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        38
           1       1.00      1.00      1.00        36
           2       1.00      1.00      1.00        32
           3       1.00      1.00      1.00        56
           4       1.00      0.97      0.98        31
           5       1.00      0.97      0.99        36
           6       1.00      1.00      1.00        34
           7       1.00      1.00      1.00        34
           8       1.00      1.00      1.00        27
           9       0.95      1.00      0.97        36

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



## 4. SGD Classifier로 예측 평가

In [21]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import classification_report

# 데이터 준비하기
digits=load_digits()
digits_data = digits.data
digits_label = digits.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data, digits_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = SGDClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.92      0.96        38
           1       0.94      0.94      0.94        36
           2       0.97      1.00      0.98        32
           3       0.98      1.00      0.99        56
           4       1.00      0.97      0.98        31
           5       0.94      0.94      0.94        36
           6       1.00      1.00      1.00        34
           7       1.00      1.00      1.00        34
           8       0.96      0.96      0.96        27
           9       0.95      1.00      0.97        36

    accuracy                           0.97       360
   macro avg       0.98      0.97      0.97       360
weighted avg       0.98      0.97      0.97       360



## 5. LogisticRegression 으로 예측 평가

In [22]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# 데이터 준비하기
digits=load_digits()
digits_data = digits.data
digits_label = digits.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data, digits_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.95      0.97        38
           1       0.95      0.97      0.96        36
           2       1.00      0.97      0.98        32
           3       1.00      0.98      0.99        56
           4       1.00      0.97      0.98        31
           5       0.94      0.92      0.93        36
           6       1.00      1.00      1.00        34
           7       1.00      0.97      0.99        34
           8       0.96      1.00      0.98        27
           9       0.88      1.00      0.94        36

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


# Project 1 digits - 평가지표 선택
### 1. DecisionTreeClassifier
- Precision의 평균 : 0.88
- Recall의 평균 : 0.88
- F1 Score : 0.88

### 2. RandomForestClassifier
- Precision의 평균 : 0.99
- Recall의 평균 : 0.98
- F1 Score : 0.99

### 3. SVM
- Precision의 평균 : 0.99
- Recall의 평균 : 0.99
- F1 Score : 0.99

### 4. SGD Classifier
- Precision의 평균 : 0.97
- Recall의 평균 : 0.96
- F1 Score : 0.96

### 5. LogisticRegression
- Precision의 평균 : 0.98
- Recall의 평균 : 0.97
- F1 Score : 0.97

### 선택
- digits 분류 문제에서는 **RandomForestClassifier, SVM**이 성능이 좋게나왔다.
- precision, recall이 높다는것은 True인 문제를 True로 올바르게 맞췄다는 의미이고, 그 둘의 지표를 균형있게 반영하여 모델의 성능을 나타내주는 F1 Score이 둘 다 0.99로 동일 하였다.

# Project 2 - wine 분류

In [23]:
# module import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [24]:
# 데이터준비
wine = load_wine()

In [25]:
# feature data 지정
wine_data = wine.data
wine_data.shape

(178, 13)

In [26]:
print(dir(wine))
print(wine.keys())

['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']
dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])


In [27]:
# label data 지정
wine_label = wine.target
wine_label.shape

(178,)

In [28]:
# target names 출력
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [29]:
# Data Describe
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

## 1. DecisionTreeClassifier

In [30]:
# module import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
wine=load_wine()
# feature data 지정
wine_data = wine.data
# label data 지정
wine_label = wine.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data, wine_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = DecisionTreeClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       0.90      0.90      0.90        10
           2       0.90      0.90      0.90        10

    accuracy                           0.94        36
   macro avg       0.93      0.93      0.93        36
weighted avg       0.94      0.94      0.94        36



## 2. RandomForestClassifier

In [31]:
# module import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
wine=load_wine()
# feature data 지정
wine_data = wine.data
# label data 지정
wine_label = wine.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data, wine_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = RandomForestClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       1.00      0.90      0.95        10
           2       0.91      1.00      0.95        10

    accuracy                           0.97        36
   macro avg       0.97      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36



## 3. SVM

In [32]:
# module import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
wine=load_wine()
# feature data 지정
wine_data = wine.data
# label data 지정
wine_label = wine.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data, wine_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = svm.SVC()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.93      0.81      0.87        16
           1       0.41      0.90      0.56        10
           2       0.00      0.00      0.00        10

    accuracy                           0.61        36
   macro avg       0.45      0.57      0.48        36
weighted avg       0.53      0.61      0.54        36



/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## 4. SGD Classifier

In [33]:
# module import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
wine=load_wine()
# feature data 지정
wine_data = wine.data
# label data 지정
wine_label = wine.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data, wine_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = SGDClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.81      0.84        16
           1       1.00      0.20      0.33        10
           2       0.42      0.80      0.55        10

    accuracy                           0.64        36
   macro avg       0.76      0.60      0.57        36
weighted avg       0.78      0.64      0.62        36



## 5. LogisticRegression

In [34]:
# module import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
wine=load_wine()
# feature data 지정
wine_data = wine.data
# label data 지정
wine_label = wine.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data, wine_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      1.00      0.97        16
           1       0.89      0.80      0.84        10
           2       0.90      0.90      0.90        10

    accuracy                           0.92        36
   macro avg       0.91      0.90      0.90        36
weighted avg       0.92      0.92      0.91        36



/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


# Project 2 wine - 평가지표 선택
### 1. DecisionTreeClassifier
- Precision의 평균 : 0.93
- Recall의 평균 : 0.93
- F1 Score : 0.93

### 2. RandomForestClassifier
- Precision의 평균 : 0.97
- Recall의 평균 : 0.97
- F1 Score : 0.97

### 3. SVM
- Precision의 평균 : 0.45
- Recall의 평균 : 0.57
- F1 Score : 0.48

### 4. SGD Classifier
- Precision의 평균 : 0.76
- Recall의 평균 : 0.60
- F1 Score : 0.57

### 5. LogisticRegression
- Precision의 평균 : 0.91
- Recall의 평균 : 0.90
- F1 Score : 0.90

### 선택
- wine 분류에서는 **RandomForestClassifier의 F1 score : 0.97로 가장 높았다.**
- precision, recall이 높다는것은 True인 문제를 True로 올바르게 맞췄다는 의미이고, 그 둘의 지표를 균형있게 반영하여 모델의 성능을 나타내주는 F1 score가 0.97로 가장 높았다.

# Project 3 - breast_cancer 여부

In [35]:
# module import
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [36]:
# 데이터준비
bc = load_breast_cancer()

In [37]:
# feature data 지정
bc_data = bc.data
bc_data.shape

(569, 30)

In [38]:
print(dir(bc))
print(bc.keys())

['DESCR', 'data', 'feature_names', 'filename', 'frame', 'target', 'target_names']
dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])


In [39]:
# label data 지정
bc_label = bc.target
bc_label.shape

(569,)

In [40]:
# target names 출력
bc.target_names

array(['malignant', 'benign'], dtype='<U9')

In [41]:
# Data Describe
print(bc.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

## 1. DecisionTreeClassifier

In [42]:
# module import
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
bc=load_breast_cancer()
# feature data 지정
bc_data = bc.data
# label data 지정
bc_label = bc.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(bc_data, bc_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = DecisionTreeClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.89      0.88        44
           1       0.93      0.91      0.92        70

    accuracy                           0.90       114
   macro avg       0.90      0.90      0.90       114
weighted avg       0.90      0.90      0.90       114



## 2. RandomForestClassifier

In [43]:
# module import
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
bc=load_breast_cancer()
# feature data 지정
bc_data = bc.data
# label data 지정
bc_label = bc.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(bc_data, bc_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = RandomForestClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.91      0.91      0.91        44
           1       0.94      0.94      0.94        70

    accuracy                           0.93       114
   macro avg       0.93      0.93      0.93       114
weighted avg       0.93      0.93      0.93       114



## 3. SVM

In [44]:
# module import
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
bc=load_breast_cancer()
# feature data 지정
bc_data = bc.data
# label data 지정
bc_label = bc.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(bc_data, bc_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = svm.SVC()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.91      0.73      0.81        44
           1       0.85      0.96      0.90        70

    accuracy                           0.87       114
   macro avg       0.88      0.84      0.85       114
weighted avg       0.87      0.87      0.86       114



## 4. SGDClassifier

In [45]:
# module import
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
bc=load_breast_cancer()
# feature data 지정
bc_data = bc.data
# label data 지정
bc_label = bc.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(bc_data, bc_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = SGDClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.47      1.00      0.64        44
           1       1.00      0.30      0.46        70

    accuracy                           0.57       114
   macro avg       0.74      0.65      0.55       114
weighted avg       0.80      0.57      0.53       114



## 5. Logistic Regression

In [46]:
# module import
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

# 데이터 준비하기
bc=load_breast_cancer()
# feature data 지정
bc_data = bc.data
# label data 지정
bc_label = bc.target

# 데이터 train_set, test_set 분리
X_train, X_test, y_train, y_test = train_test_split(bc_data, bc_label, test_size=0.2, random_state=32)

# 모델 학습 및 예측
model = LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.89      0.89      0.89        44
           1       0.93      0.93      0.93        70

    accuracy                           0.91       114
   macro avg       0.91      0.91      0.91       114
weighted avg       0.91      0.91      0.91       114



/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


# Project 3 brest_cancer - 평가지표 선택
### 1. DecisionTreeClassifier
- Precision의 평균 : 0.90
- Recall의 평균 : 0.90
- F1 Score : 0.90

### 2. RandomForestClassifier
- Precision의 평균 : 0.93
- Recall의 평균 : 0.93
- F1 Score : 0.93

### 3. SVM
- Precision의 평균 : 0.88
- Recall의 평균 : 0.84
- F1 Score : 0.85

### 4. SGD Classifier
- Precision의 평균 : 0.74
- Recall의 평균 : 0.65
- F1 Score : 0.55

### 5. LogisticRegression
- Precision의 평균 : 0.91
- Recall의 평균 : 0.91
- F1 Score : 0.91

### 선택
- breast_cancer 분류에서는 **RandomForestClassifier의 F1 score : 0.93으로 가장 높았다.**
- precision, recall이 높다는것은 True인 문제를 True로 올바르게 맞췄다는 의미이고, 그 둘의 지표를 균형있게 반영하여 모델의 성능을 나타내주는 F1 score가 0.93로 가장 높았다.